### PR TITLE
Cart Cleanup Job

### PR DESCRIPTION
**Pull request by andela-aatanda**
Resolves [1411](https://github.com/reactioncommerce/reaction/issues/1411)

**Test**
- Add `"cart": { "cleanupDurationDays": "older than 3 days" }` to "settings" in `reaction.json`

***Anonymous User***
- Pull this branch to your local machine
- Launch the application and put an item in the cart without login and checkout attempt.

***Registered User***
- Pull this branch to your local machine
- Launch application and Login with your credentials
- Select a product and place in your cart without checking out

**Note:**
You'll have to reset the schedule in `imports/plugins/included/jobcontrol/jobs/cart.js` as current schedule is daily

### CODE REVIEW

File: imports/plugins/core/dashboard/client/templates/shop/settings/settings.html [hunk 1]
```diff
@@ -102,6 +102,7 @@
             {{> afQuickField name='settings.openexchangerates.refreshPeriod' placeholder="every 1 hour"}}
             {{> afQuickField name='settings.google.clientId'}}
             {{> afQuickField name='settings.google.apiKey'}}
+            {{> afFormGroup name='settings.cart.cleanup' options=options placeholder="Every 3 days"}}
```
Comment (zenweasel) [inline imports/plugins/core/dashboard/client/templates/shop/settings/settings.html]: This should just be a text field. Also the name should be more descriptive like `cleanupDurationDays`

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
+import later from "later";
+import moment from "moment";
+import { Accounts, Cart, Jobs } from "/lib/collections";
+import { Hooks, Logger, Reaction } from "/server/api";
+
+const olderThan = moment().subtract(3, "days")._d;
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: You aren't using the settings variable you created.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        backoff: "exponential" // delay by twice as long for each subsequent retry
+      })
+      .repeat({
+        schedule: later.parse.text("every day")
+      })
+      .save({
+        cancelRepeats: true
+      });
+  } else {
+    Logger.info("No stale carts at the moment");
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: This shouldn't be an `info` level log. We don't need to log things we **don't** do.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            Logger.debug(success);
+            job.done(success, { repeatId: true });
+          }
+        } else {
+          const settings = Reaction.getShopSettings();
+          const schedule = Number(settings.cart.cleanup); // configurable in shop settings
+          if (!schedule) {
+            Logger.warn("Cleanup Schedule not configured");
+          } else {
+            if (cart.updatedAt <= moment().subtract(schedule, "mins")._d) {
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: I don't understand what you are doing here? Removing all items from the cart?

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      updatedAt: {
+        $lte: olderThan
+      }
+    }).fetch();
+    cartDetails.map(cart => {
+      if (cart.items) {
+        const user = Accounts.findOne({ userId: cart.userId });
+        if (!user.emails.length) {
+          const result = Cart.update(
+            { userId: user._id },
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: By "cleaning the cart" I think we meant removing the cart record. We should also be removing stale Account and Session records as noted in the ticket.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+          }
+        }
+      } else {
+        Logger.debug("Cart is currently empty");
+      }
+    });
+    callback();
+  });
+  Jobs.find({
+    type: "cart/removeFromCart",
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: What is this supposed to be doing?

File: imports/plugins/included/jobcontrol/server/jobs/cart.js (hunk 4 above)
Comment (andela-aatanda) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: Yeah, I'm removing all items from the cart here. The `mins` on line 64 should have been `days`.

Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: My point being, removing items from the cart is not what we want. We want to remove all unused/unneeded cart records.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js (hunk 3 above)
Comment (andela-aatanda) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: Noted.

File: imports/plugins/core/dashboard/client/templates/shop/settings/settings.html (hunk 1 above)
Comment (andela-aatanda) [inline imports/plugins/core/dashboard/client/templates/shop/settings/settings.html]: Noted.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+              const success = "Stale anonymous user cart and account successfully cleaned";
+              Logger.debug(success);
+              job.done(success, { repeatId: true });
+            }
+          } else {
+            const user = Accounts.findOne({ _id: cart.userId });
+            if (cart.updatedAt <= moment().subtract(Number(schedule[0]), "days")._d) {
+              const userCleanup = Cart.update(
+                { userId: user._id },
+                { $unset: { items: [] } }
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: I'm still unclear why we are ever removing items from the cart.

**Comment by zenweasel** [comment]
When I start this after `reaction reset` I get the message ` WARN Reaction: No cart cleanup schedule` and nothing is added.

File: imports/plugins/core/dashboard/client/templates/shop/settings/settings.html [hunk 8]
```diff
@@ -102,6 +102,7 @@
             {{> afQuickField name='settings.openexchangerates.refreshPeriod' placeholder="every 1 hour"}}
             {{> afQuickField name='settings.google.clientId'}}
             {{> afQuickField name='settings.google.apiKey'}}
+            {{> afQuickField name='settings.cart.cleanupDurationDays' placeholder="Every 3 days"}}
```
Comment (zenweasel) [inline imports/plugins/core/dashboard/client/templates/shop/settings/settings.html]: Unless I am misunderstanding this is a duration, but it's stated as a schedule i.e. "every 3 days". This job runs once a day regardless correct? So this should be "Older than X days" or something similar.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        if (cart.items) {
+          if (anonUser) {
+            const removeCart = Cart.remove({ userId: anonUser._id });
+            const removeAccount = Accounts.remove(
+              {
+                _id: cart.userId,
+                emails: []
+              }
+            );
+            const destroySession = ServerSessions.remove({ _id: cart.sessionId });
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: This logic is a little hard to follow and this is hard to test (not your fault, it's just the nature of these jobs).

Can we break this out into a function that could be called at the command line e.g. `purgeAnonymousCarts(3)` that would purge Anonymous carts older than 3 days (or whatever argument is passed there)? Even better if we could specify time unit, e.g. "1 day" or "2 hours". Then I would feel better about being able to test that it does what it says it's doing as I could just call it from the meteor shell.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        } else {
+          Logger.debug("No item in cart");
+        }
+      });
+    } else {
+      Logger.warn("No cart cleanup schedule");
+    }
+    callback();
+  });
+  Jobs.find({
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: Can you help me understand how this is supposed to work?

**Comment by andela-aatanda** [comment]
That's correct @zenweasel You have to setup cleanup schedule on the dashboard shop `options` before the job can be added.

**Comment by zenweasel** [comment]
@andela-aatanda Yes, I did that. It doesn't work

File: imports/plugins/included/jobcontrol/server/jobs/cart.js (hunk 7 above)
Comment (andela-aatanda) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: If the entire cart of a user is removed, there's a brief disappearance of the cart icon on the landing page until there's a reload and I don't think that is a wanted behavior which was what informed my decision to remove items instead.

File: imports/plugins/core/dashboard/client/templates/shop/settings/settings.html (hunk 8 above)
Comment (andela-aatanda) [inline imports/plugins/core/dashboard/client/templates/shop/settings/settings.html]: Yeah. It's misleading. It has been corrected

File: imports/plugins/included/jobcontrol/server/jobs/cart.js (hunk 7 above)
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: But if this job is working correctly, when would you ever be eliminating the cart for a currently active user?

Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: Just to be clear, the entire point of this job is that lots of cart records increases seek time. So just removing items from the cart would not going with what this code should be trying to accomplish.

Comment (andela-aatanda) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: The assumption here is that the user has items that are older than a certain number of days and does not check if the user is currently logged in. Hence, the user may or may not be logged in when this job runs.

Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: So the use case you are talking about is of a user who goes to the site, places something in the cart, and then stares at the cart icon for three days to see it briefly disappear?

Comment (andela-aatanda) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: lol. That's correct

File: imports/plugins/included/jobcontrol/server/jobs/cart.js (hunk 10 above)
Comment (andela-aatanda) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: This is supposed to be an observer that triggers a job after it has been completed. I see it's that way for currently existing jobs

File: imports/plugins/included/jobcontrol/server/jobs/cart.js (hunk 9 above)
Comment (andela-aatanda) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: This has been implemented on line 35.

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    Logger.warn("No cart cleanup schedule");
+  }
+});
+
+/**
+ * {Function} that fetches stale carts
+ * @param {Object} olderThan older than date
+ * @return {Object} stale carts
+ */
+const staleCarts = (olderThan) => {
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: this function should be called `getStaleCarts`. Function names should always contain verbs

File: imports/plugins/included/jobcontrol/server/jobs/cart.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    workTimeout: 180 * 1000
+  }, (job, callback) => {
+    Logger.debug("Processing cart/removeFromCart");
+    const settings = Reaction.getShopSettings();
+    if (settings.cart) {
+      const schedule = (settings.cart.cleanupDurationDays).match(/\d/);// configurable in shop settings
+      const olderThan = moment().subtract(Number(schedule[0]), "days")._d;
+      const carts = staleCarts(olderThan);
+      carts.forEach(cart => {
+        if (cart.items) {
```
Comment (zenweasel) [inline imports/plugins/included/jobcontrol/server/jobs/cart.js]: So if I am reading this right, if the cart is empty we aren't deleting it? That definitely not what we want to do.
